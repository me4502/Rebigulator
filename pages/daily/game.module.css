.scoreBox {
  font-family: 'akbarplain';
  padding: 0rem 1rem 0 1rem;
  text-align: center;
  font-size: 20pt;
}

.loadingBox {
  display: flex;
  align-items: center;
  flex-direction: column;
}
