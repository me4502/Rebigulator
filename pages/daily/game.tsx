import type { FC } from 'react';
import { ToastContainer } from 'react-toastify';
import Layout from '../../src/components/layout';
import SEO from '../../src/components/seo';

const GamePage: FC = () => {
  return (
    <Layout>
      <SEO title="Play the daily challenge | The Rebigulator" />
      <ToastContainer draggable={true} closeOnClick={true} />
    </Layout>
  );
};

export default GamePage;
