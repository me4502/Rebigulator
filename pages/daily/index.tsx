import type { FC } from 'react';

import Layout from '../../src/components/layout';
import SEO from '../../src/components/seo';
import { Container } from '../../src/components/Container.module.css';
import { MainButtonLink } from '../../src/components/MainLink';
import BrandSvg from '../../src/images/brand.svg';
import { bannerImage, centreDiv } from '../index.module.css';

const IndexPage: FC = () => (
  <Layout>
    <SEO title="The Rebigulator Daily Mode" />
    <div className={Container}>
      <BrandSvg className={bannerImage} />
      <div className={centreDiv}>
        <h1>
          The Rebigulator Daily Mode is a daily online fast-paced quote-based
          Simpsons Trivia Quiz Game, with a new challenge every day!
        </h1>
        <h2 style={{ marginBottom: '3rem' }}>
          Guess the daily episode from the shown frame. Each wrong guess you
          make will show you a new frame from the episode.
        </h2>
        <MainButtonLink href="/daily/game/">Get started now!</MainButtonLink>
      </div>
    </div>
  </Layout>
);

export default IndexPage;
