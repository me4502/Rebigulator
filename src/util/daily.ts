import { useFrinkiacEpisodes } from '../frinkiac/episodes';
import type { FrinkiacEpisodesJsonType } from '../frinkiac/types';

export const ROUND_COUNT = 6;

/**
 * Gets the current day's date, to use as a seed for daily challenges.
 * This is less perfect than manually setting the challenges each day,
 * but it means that I don't have to set a list of challenges every day.
 *
 * @returns A string representing today's date in YYYY-MM-DD format.
 */
export const getDateString = (): string => {
  const today = new Date();
  return today.toISOString().split('T')[0]; // YYYY-MM-DD format
};

export function getDailyEpisode(
  dateString: string,
  episodes: FrinkiacEpisodesJsonType[]
): FrinkiacEpisodesJsonType {
  // Simple hash function to convert date string to number
  let hash = 0;
  for (let i = 0; i < dateString.length; i++) {
    const char = dateString.codePointAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash; // Convert to 32-bit integer
  }

  // Use absolute value and modulo to get episode index
  const episodeIndex = Math.abs(hash) % episodes.length;
  return episodes[episodeIndex];
}

export function useDailyEpisode(date?: string): FrinkiacEpisodesJsonType {
  const dateString = date || getDateString();

  return useFrinkiacEpisodes((data) => {
    return getDailyEpisode(dateString, data);
  });
}

export function getDailyTimestampHashes(date?: string): number[] {
  const dateString = date || getDateString();

  const timestampHashes: number[] = [];

  for (let i = 0; i < ROUND_COUNT; i++) {
    // Create very different seeds using large prime multipliers and different salts
    const seedString = `${dateString}-round${i * 12347}-salt${i * 98765 + 54321}`;
    let hash = 5381; // djb2 algorithm starting value

    // Hash the seed string
    for (let j = 0; j < seedString.length; j++) {
      const char = seedString.codePointAt(j) || 0;
      hash = (hash << 5) + hash + char; // hash * 33 + char
      hash = hash >>> 0; // Convert to unsigned 32-bit
    }

    // Apply additional mixing with different constants per iteration
    const mixingConstant = [
      0x85ebca6b, 0xc2b2ae35, 0xcc9e2d51, 0x1b873593, 0xe6546b64, 0x9e3779b9,
    ][i % 6];
    hash ^= hash >>> 16;
    hash *= mixingConstant;
    hash ^= hash >>> 13;
    hash *= 0xc2b2ae35;
    hash ^= hash >>> 16;
    hash = hash >>> 0;

    // Additional round-specific mixing
    hash ^= (i * 2654435761) >>> 0; // Different prime per round
    hash = hash >>> 0;

    // Normalize to 0-1 range
    const normalizedHash = hash / 4294967295; // Max unsigned 32-bit value
    timestampHashes.push(normalizedHash);
  }

  return timestampHashes;
}

export interface DailyResults {
  a: number; // Attempts
  r: (string | null)[]; // Results
  e: string; // Episode Key
  w: boolean; // Won
  d?: string; // Date
}

export function generateDailyResults(
  attempts: number,
  results: (string | null)[],
  episodeKey: string,
  won: boolean,
  date?: string
): string {
  return `${btoa(JSON.stringify({ a: attempts, r: results, e: episodeKey, w: won, d: date }))}`;
}
