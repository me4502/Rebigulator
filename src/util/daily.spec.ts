import { describe, expect, it } from 'vitest';
import { getDailyEpisode, getDailyTimestampHashes } from './daily';

describe('getDailyEpisode', () => {
  it.each([['2023-03-01'], ['2024-10-01'], ['2025-07-03']])(
    'should return the correct episode for the given date %s',
    (dateString) => {
      const date = new Date(dateString).toISOString().split('T')[0];
      const episode = getDailyEpisode(date);
      expect(episode).toMatchSnapshot();
    }
  );
});

describe('getDailyTimestampHashes', () => {
  it.each([['2023-03-01'], ['2024-10-01'], ['2025-07-03']])(
    'should return the correct timestamp hashes for the given date %s',
    (dateString) => {
      const date = new Date(dateString).toISOString().split('T')[0];
      const episode = getDailyTimestampHashes(date);
      expect(episode).toMatchSnapshot();
    }
  );
});
