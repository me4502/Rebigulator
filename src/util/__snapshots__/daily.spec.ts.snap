// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDailyEpisode > should return the correct episode for the given date 2023-03-01 1`] = `
{
  "label": "The Boys of Bummer",
  "value": "S18E18",
}
`;

exports[`getDailyEpisode > should return the correct episode for the given date 2024-10-01 1`] = `
{
  "label": "The City of New York vs. Homer Simpson",
  "value": "S09E01",
}
`;

exports[`getDailyEpisode > should return the correct episode for the given date 2025-07-03 1`] = `
{
  "label": "Homer the Heretic",
  "value": "S04E03",
}
`;

exports[`getDailyTimestampHashes > should return the correct timestamp hashes for the given date 2023-03-01 1`] = `
[
  0.5293715308721577,
  0.702072634757979,
  0.7895707892695374,
  0.8930267786823741,
  0.332600548009528,
  0.6933471038223586,
]
`;

exports[`getDailyTimestampHashes > should return the correct timestamp hashes for the given date 2024-10-01 1`] = `
[
  0.6646155204308721,
  0.7453723751812643,
  0.6031946778770524,
  0.7609660317564769,
  0.1532445629484124,
  0.48569006996361774,
]
`;

exports[`getDailyTimestampHashes > should return the correct timestamp hashes for the given date 2025-07-03 1`] = `
[
  0.37401850506989714,
  0.7887633095469241,
  0.7878796306410524,
  0.774008486600129,
  0.4678007272229997,
  0.5573874601529416,
]
`;
